linkerd/app/gateway/src/lib.rs

```diff
@@ -162,12 +179,19 @@ where
     // The client's ID is set as a request extension, as required by the
     // gateway. This permits gateway services (and profile resolutions) to be
     // cached per target, shared across clients.
-    let http = outbound
-        .push_tcp_endpoint()
-        .push_http_endpoint()
+    let endpoint = outbound.push_tcp_endpoint().push_http_endpoint();
+    let http = endpoint
+        .clone()
         .push_http_logical(resolve)
         .into_stack()
-        .push(NewGateway::layer(local_id))
+        .push_switch(
+            |t: svc::Either<outbound::http::Logical, outbound::http::Endpoint>| -> Result<
+                svc::Either<outbound::http::Logical, outbound::http::Endpoint>,
+                Infallible,
+            > { Ok(t) },
+            endpoint.into_stack(),
+        )
```

Comment:
It turns out this can be simplified quite a bit as:

```suggestion
        .push_switch(Ok::<_, Never>, endpoint.into_stack())
```

First, I tried to remove the return signature by instead encoding the error type as `Ok::<_, Never>(t)`. (It's fine to use `Infallible` in general, but we already have `Never` in scope -- we could probably get rid of our `Never` type in favor of `Infallible` but that's a separate change). We can know the success value based on `t` -- but since the error variant is never constructed we need to explicitly annotate it.

Then, clippy gives me the warning:
```
error: redundant closure
   --> linkerd/app/gateway/src/lib.rs:188:13
    |
188 |             |t: svc::Either<outbound::http::Logical, outbound::http::Endpoint>| Ok::<_, Never>(t),
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `Ok::<_, Never>`
    |
note: the lint level is defined here
```

Which is saying that `Ok::<_, Never>` is equivalent to 
```rust
|t: svc::Either<outbound::http::Logical, outbound::http::Endpoint>| Ok::<_, Never>(t)
```

Reply:
This looks much cleaner, made the change. I went ahead and used `Never` as an error type here since it was already exported. I wasn't sure if I should pull in some of the changes from #1158 or wait until it is merged and then rebase off the latest main. Let me know what would work better here.

Reply:
I wouldn't merge it in until it's on main. I was going to try to let this merge first to reduce overhead for you, or if the other gets merged first I'd probably merge main into this branch for you